### DISCUSSION TITLE
Composite pseudo-states

### DISCUSSION THREAD

**Discussion by jmagaram** [statelyai/xstate#2006] [Q&A] [answered]
I've got a situation where there are three states ABC and all of these states should handle event X identically. I can repeat the event transition 3 times - once for each source state - and obviously this is not ideal. I can group ABC into a new state S and define the event transition there once. But the problem is that state S is a pseudo-state - introduced to reduce the number of transitions - and does not have an initial/default state like most composite states. There will never be a transition to state S and instead any transition will go directly to a child state. Is it possible to do this with xState? I tried using the `on` syntax directly as a child of `states` but it didn't seem to work. I read Ian Horrock's book a while ago about coding the UI with state charts and remember this being a useful feature. I've got a workaround for my particular case but am curious if xState supports this kind of grouping states into composite pseudo-states.

**Comment by davidkpiano** [comment, marked as answer]
> There will never be a transition to state S and instead any transition will go directly to a child state. Is it possible to do this with xState?

Yes, sorta... the workaround (right now) is to define a default `initial` state on the parent state, even if it's not needed:

```js
S: {
  initial: 'A', // doesn't matter
  on: {
    EVENT: { actions: () => {/* common action */} }
  },
  states: {
    A: {},
    B: {},
    C: {}
  }
}
```

Of course, you can also do this in "plain-JS-land":

```js
const commonTransition = { EVENT: {/* ... */} };

// ...

states: {
  A: {
    on: { ...commonTransition }
  },
  B: {
    on: { ...commonTransition }
  },
  C: {
    on: { ...commonTransition }
  }
}
```
